padertorch/data/fragmenter.py

```diff
@@ -109,86 +119,145 @@ def fragment(key, x):
-        x:
-        segment_length:
-        step:
-        axis:
-        offset_mode:
-        num_segments:
+        x: signal to be segmented
+        axis: axis which is segmented
+        boundary: 2xB numpy array with start and end values
+            for B pre calculated boundaries, which may be used to segment
+            multiple audios in the same way.
+        length: segment length
+        shift: shift between segments
+        mode: Defines the position of the boundaries in the signal:
+            begin: boundaries start at sample zero
+                so that only values at the end are cut
+            end: boundaries end at num_samples
+                so that only values at the beginning are cut
+            center: boundaries are chosen such that the same number of samples
+                are discarded at the end and the beginning of the signal
+            random: Starts the first segment randomly at a value
+                between 0 and shift
+            random_max_segments: Randomly chooses the first segment such that
+                the maximum number of segments are created
+        choose_one: If True only outputs one segment boundary
+        rng: random number generator (numpy.random)
 
     Returns:
 
-    >>> segment(np.arange(0, 15), 10, None, 3, -1)
+    >>> segment(np.arange(0, 15), -1, None, 10, 3)
     array([[ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9],
            [ 3,  4,  5,  6,  7,  8,  9, 10, 11, 12]])
     """
 
-    if offset is None:
-        offset = get_segment_offsets(x.shape[axis], length, shift,
-                                      offset_mode, num_segments)
+    if boundary is None:
+        boundary = get_segment_boundaries(x.shape[axis], length, shift,
+                                            mode, choose_one, rng)
+    else:
+        assert length is None, (
+            'Either use predefined boundaries or set a segment length'
+            'because only one of these information can be used.'
+        )
+        if isinstance(boundary, (list, tuple)):
+            boundary = np.array(boundary)
+            if boundary.ndim == 1:
+                boundary = boundary[..., None]
+            assert boundary.ndim == 2, boundary.shape
+        elif x.__class__.__module__ == 'numpy':
+            assert x.ndim == 2, x.shape
+        elif x.__class__.__module__ == 'torch':
+            assert x.dim() == 2, x.shape
+        else:
+            raise TypeError('The signal x has an unknown type', type(x))
+
+        assert boundary.shape[0] == 2, 'The first dimension has to be 2 to' \
+                                       ' describe start and stop of a segment'
+
     # slice the array to remove samples not addressed with
     # this offsets-segment_length combination
     slc = [slice(None)] * x.ndim
-    slc[axis] = slice(offset[0], offset[-1] + length)
+    slc[axis] = slice(boundary[0, 0], boundary[1, -1])
     x = x[tuple(slc)]
 
     return segment_axis(x, length, shift, end='cut')
```

Comment:
This only works if `boundaries` has the correct format. If you pass in custom boundaries that don't match up with a constant integer shift and size, this gives unexpected results. Can we also support passing in boundaries that don't use a constant shift and size?

Reply:
good idea, I will replace this with a max and min operator

Reply:
I think it would be better to drop the end value of boundary and only use `begin` as argument. That is enough and much easier for the user.

Reply:
If you keep this signature, a user may think, this function can do much more fancy things.

Reply:
hmm, okay I have gone a different way and have added the possiblity to use more possible inputs but you are right I will allow the user to input a one dimensional array and in that case the end part of the array will be added by using a fixed length if it was specified.

Reply:
support boundaries that don't use a constant shift and size would be cool. How about inferring shifts and lengths from the boundary array? If they are the same for all segments use segment_axis and else a for loop?

Reply:
I think there are a few signatures a user may expect:
 - `segment(x, length, shift, mode) -> array`
 - `segment(x, length, shift, begin, end=None) -> array`
 - `segment(x, boundaries) -> list[array]`

When a user specifies the `boundaries` argument, `length` and `shift` should be ignored and the code uses a loopy implementation (i.e. return type is list of array).

When `boundaries` is not given, the return type can be a multidimensional array and `segment_axis` should be used.

To force a user that want to use beginning has to specify boundaries is not a good idea (Too much boilerplate).

I don't like the idea to use magic to figure out if boundaries has common size and shifts.
It can randomly change the return type.